### PR TITLE
[lldb-dap] Support throw and catch exception breakpoints for dynamica…

### PR DESCRIPTION
**Pull request by walter-erquinigo**
…lly registered languages

First of all, this is done to support exceptions for the Mojo language, but it's done in a way that will benefit any other plugin language.

1. I added a new lldb-dap CLI argument (not DAP field) called `pre-init-commands`. These commands are executed before DAP initialization. The other `init-commands` are executed after DAP initialization. It's worth mentioning that the debug adapter returns to VSCode the list of supported exception breakpoints during DAP initialization, which means that I need to register the Mojo plugin before that initialization step, hence the need for `pre-init-commands`. In general, language plugins should be registered in that step, as they affect the capabilities of the debugger.
2. I added a set of APIs for lldb-dap to query information of each language related to exception breakpoints. E.g. whether a language supports throw or catch breakpoints, how the throw keyword is called in each particular language, etc.
3. I'm realizing that the Swift support for exception breakpoints in lldb-dap should have been implemented in this way, instead of hardcoding it.

### CODE REVIEW

**Comment by walter-erquinigo** [comment]
This is a screenshot of how it's looking for me on VSCode.
image

File: lldb/source/API/SBLanguageRuntime.cpp [hunk 1]
```diff
@@ -26,3 +26,43 @@ SBLanguageRuntime::GetNameForLanguageType(lldb::LanguageType language) {
...
+    lldb::LanguageType language) {
+  if (Language *lang_plugin = Language::FindPlugin(language))
+    return lang_plugin->SupportsExceptionBreakpointsOnCatch();
+  return false;
+}
+
+const char *
+SBLanguageRuntime::GetThrowKeywordForLanguage(lldb::LanguageType language) {
+  if (Language *lang_plugin = Language::FindPlugin(language))
+    return lang_plugin->GetThrowKeyword().data();
```
Comment (bulbazord) [inline lldb/source/API/SBLanguageRuntime.cpp:59]: You should wrap these in `ConstString`. It is not safe in general for LLDB to pass out `const char *` values that are derived from `llvm::StringRef`. I know that `GetThrowKeyword` and `GetCatchKeyword` are passing out static c-strings wrapped in a `llvm::StringRef`, but those are virtual and plugins may not do the same.

File: lldb/tools/lldb-dap/DAP.cpp [hunk 2]
```diff
@@ -58,10 +58,17 @@ DAP::DAP()
...
+  if (str.empty())
+    return "";
+  return ((llvm::Twine)llvm::toUpper(str[0]) + str.drop_front()).str();
+}
+
 void DAP::PopulateExceptionBreakpoints() {
   llvm::call_once(init_exception_breakpoints_flag, [this]() {
     exception_breakpoints = std::vector<ExceptionBreakpoint> {};
-    
+
```
Comment (bulbazord) [inline lldb/tools/lldb-dap/DAP.cpp:71]: Unrelated whitespace change

File: lldb/tools/lldb-dap/DAP.cpp [hunk 3]
```diff
@@ -80,6 +87,46 @@ void DAP::PopulateExceptionBreakpoints() {
...
+      std::string throw_keyword =
+          raw_throw_keyword ? raw_throw_keyword : "throw";
+      std::string catch_keyword =
+          raw_catch_keyword ? raw_catch_keyword : "catch";
+
+      if (lldb::SBLanguageRuntime::SupportsExceptionBreakpointsOnThrow(lang)) {
+        exception_breakpoints->emplace_back(
+            raw_lang_name + "_" + throw_keyword,
+            capitalized_lang_name + " " + capitalize(throw_keyword), lang);
+      }
```
Comment (bulbazord) [inline lldb/tools/lldb-dap/DAP.cpp:123]: nit: [single-statement if ](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)

File: lldb/tools/lldb-dap/DAP.cpp [hunk 4]
```diff
@@ -80,6 +87,46 @@ void DAP::PopulateExceptionBreakpoints() {
...
+      if (lldb::SBLanguageRuntime::SupportsExceptionBreakpointsOnCatch(lang)) {
+        exception_breakpoints->emplace_back(
+            raw_lang_name + "_" + catch_keyword,
+            capitalized_lang_name + " " + capitalize(catch_keyword), lang);
+      }
```
Comment (bulbazord) [inline lldb/tools/lldb-dap/DAP.cpp:128]: same as above

File: lldb/tools/lldb-dap/DAP.cpp (hunk 2 above)
Comment (walter-erquinigo) [reply lldb/tools/lldb-dap/DAP.cpp:71]: This one doesn't affect git blame history. It should be fine :)

File: lldb/include/lldb/API/SBLanguageRuntime.h [hunk 5]
```diff
@@ -18,6 +18,32 @@ class SBLanguageRuntime {
   static lldb::LanguageType GetLanguageTypeFromString(const char *string);
 
   static const char *GetNameForLanguageType(lldb::LanguageType language);
+
+  /// Returns whether the given language is any version of  C++.
```
Comment (bulbazord) [inline lldb/include/lldb/API/SBLanguageRuntime.h:22]: nit: extra space between `of` and `C++` at the end.

File: lldb/source/API/SBLanguageRuntime.cpp [hunk 6]
```diff
@@ -26,3 +26,43 @@ SBLanguageRuntime::GetNameForLanguageType(lldb::LanguageType language) {
...
+    lldb::LanguageType language) {
+  if (Language *lang_plugin = Language::FindPlugin(language))
+    return lang_plugin->SupportsExceptionBreakpointsOnCatch();
+  return false;
+}
+
+const char *
+SBLanguageRuntime::GetThrowKeywordForLanguage(lldb::LanguageType language) {
+  if (Language *lang_plugin = Language::FindPlugin(language))
+    return ConstString(lang_plugin->GetThrowKeyword().data()).AsCString();
```
Comment (bulbazord) [inline lldb/source/API/SBLanguageRuntime.cpp:59]: Shouldn't need to do `.data()` at the end there. `ConstString::ConstString` can accept `llvm::StringRef` as an argument, and even better it doesn't have to re-calculate the length of the string when you do so.

File: lldb/tools/lldb-dap/DAP.cpp [hunk 7]
```diff
@@ -80,6 +87,45 @@ void DAP::PopulateExceptionBreakpoints() {
...
+      std::string raw_lang_name = name;
+      std::string capitalized_lang_name = capitalize(name);
+      const char *raw_throw_keyword =
+          lldb::SBLanguageRuntime::GetThrowKeywordForLanguage(lang);
+      const char *raw_catch_keyword =
+          lldb::SBLanguageRuntime::GetCatchKeywordForLanguage(lang);
+      std::string throw_keyword =
+          raw_throw_keyword ? raw_throw_keyword : "throw";
+      std::string catch_keyword =
+          raw_catch_keyword ? raw_catch_keyword : "catch";
```
Comment (bulbazord) [inline lldb/tools/lldb-dap/DAP.cpp:117]: Suggestion: Instead of always calculating the keywords, why not calculate them conditionally below?

**Review by bulbazord — COMMENTED** [review]
Seems okay to me, but I'm not exactly an expert when it comes to lldb-dap. You'll probably want to get sign-off from at least another person.

**Comment by walter-erquinigo** [comment]
This is a very benign change, so I'll be merging it. Happy to make change post-merge if anyone provides any feedback.
